### ISSUE TITLE
Windows problems libuv/extpoll

### ISSUE THREAD

**Issue by lucasterra** [warmcat/libwebsockets#496]
**Compiler:** Microsoft Visual Studio Community 2015
**libwebsockets extra options:** openssl and libuv (ssl is disabled, though)
**libuv version**: 1.9.1 (latest at the time of this post)

_test-server-ext-poll_ doesn't work at all. It starts and errors at [poll](https://github.com/warmcat/libwebsockets/blob/master/test-server/test-server.c#L429). WSAGetLastError() returns code 10022 - WSAEINVAL

_test-server-libuv_ starts just fine, but then it raises an assert when a new client connects in and later on it crashes. This assert comes from libuv's [uv__get_osfhandle](https://github.com/libuv/libuv/blob/v1.x/src/win/handle-inl.h#L165-L177):
```c++
INLINE static HANDLE uv__get_osfhandle(int fd)
{
  /* _get_osfhandle() raises an assert in debug builds if the FD is invalid. */
  /* But  it also correctly checks the FD and returns INVALID_HANDLE_VALUE */
  /* for invalid FDs in release builds (or if you let the assert continue).  */
  /* So this wrapper function disables asserts when calling _get_osfhandle. */

  HANDLE handle;
  UV_BEGIN_DISABLE_CRT_ASSERT();
  handle = (HANDLE) _get_osfhandle(fd);
  UV_END_DISABLE_CRT_ASSERT();
  return handle;
}
```

Debug logs for test-server-libuv:
```
[2016/03/18 17:02:31:1423] NOTICE: libwebsockets test server libuv - license LGPL2.1+SLE
[2016/03/18 17:02:31:1423] NOTICE: (C) Copyright 2010-2016 Andy Green <andy@warmcat.com>
[2016/03/18 17:02:31:1428] NOTICE: Initial logging level 20
[2016/03/18 17:02:31:1428] NOTICE: Libwebsockets version: 1.7.0
[2016/03/18 17:02:31:1428] NOTICE: IPV6 not compiled in
[2016/03/18 17:02:31:1428] NOTICE: libev support not compiled in
[2016/03/18 17:02:31:1428] NOTICE: libuv support compiled in and enabled
[2016/03/18 17:02:31:1438] NOTICE:  Threads: 1 each 30000 fds
[2016/03/18 17:02:31:1453] NOTICE:  Compiled with OpenSSL support
[2016/03/18 17:02:31:1453] NOTICE:  SSL disabled: no LWS_SERVER_OPTION_DO_SSL_GLOBAL_INIT
[2016/03/18 17:02:31:1458] NOTICE: Creating Vhost 'default' port 7681, 4 protocols
[2016/03/18 17:02:31:1483] DEBUG: insert_wsi_socket_into_fds: 02738598: tsi=0, sock=164, pos-in-fds=0
[2016/03/18 17:02:31:1483] NOTICE:  Listening on port 7681
[2016/03/18 17:02:31:1483] NOTICE:  mem: per-conn:          632 bytes + protocol rx buf
[2016/03/18 17:02:31:1689] NOTICE:  canonical_hostname = DESKTOP-OVG8TVM
[2016/03/18 17:02:32:1709] DEBUG: lws_uv_timeout_cb
[2016/03/18 17:02:32:2040] DEBUG: fd=164, revents=9
[2016/03/18 17:02:32:2045] DEBUG: accepted new conn  port 40942 on fd=456
[2016/03/18 17:02:32:2045] DEBUG: lws_set_timeout: 024CF980: 5 secs
[2016/03/18 17:02:32:2045] DEBUG: lws_libuv_accept: new wsi 024CF980
[2016/03/18 17:02:34:6189] DEBUG: insert_wsi_socket_into_fds: 024CF980: tsi=0, sock=456, pos-in-fds=1
[2016/03/18 17:02:34:6194] DEBUG: accepted new conn  port 40943 on fd=464
[2016/03/18 17:02:34:6194] DEBUG: lws_set_timeout: 024CFE30: 5 secs
[2016/03/18 17:02:34:6194] DEBUG: lws_libuv_accept: new wsi 024CFE30
[2016/03/18 17:02:35:0730] DEBUG: insert_wsi_socket_into_fds: 024CFE30: tsi=0, sock=464, pos-in-fds=2
[2016/03/18 17:02:35:0730] DEBUG: accepted new conn  port 40944 on fd=476
[2016/03/18 17:02:35:0735] DEBUG: lws_set_timeout: 024D09D8: 5 secs
[2016/03/18 17:02:35:0735] DEBUG: lws_libuv_accept: new wsi 024D09D8
[2016/03/18 17:02:35:5278] DEBUG: insert_wsi_socket_into_fds: 024D09D8: tsi=0, sock=476, pos-in-fds=3
[2016/03/18 17:02:35:5278] DEBUG: lws_uv_idle
[2016/03/18 17:02:35:5278] DEBUG: lws_uv_idle: done stop
[2016/03/18 17:02:35:5512] DEBUG: lws_uv_timeout_cb
[2016/03/18 17:02:36:5514] DEBUG: lws_uv_timeout_cb
[2016/03/18 17:02:37:5516] DEBUG: lws_uv_timeout_cb
[2016/03/18 17:02:38:5514] DEBUG: lws_uv_timeout_cb
[2016/03/18 17:02:38:5514] NOTICE: wsi 024CF980: TIMEDOUT WAITING on 3 (did hdr 0, ah 00000000, wl 0, pfd events 0)
[2016/03/18 17:02:38:5529] DEBUG: lws_libuv_io: wsi: 024CF980, flags:0x8000000b
Assertion failed: handle->type == UV_POLL, file src\win\poll.c, line 230
```

**Comment by lws-team** [comment]
I think we might have met this already, on windows an fd can be two different types somehow.  If you pass libuv the wrong type it will blow up.  Can we see the backtrace to the assert?

**Comment by lucasterra** [comment]
```
>	libwebsockets-test-server-libuv.exe!uv__get_osfhandle(int fd) Line 174	C
 	libwebsockets-test-server-libuv.exe!uv_poll_init(uv_loop_s * loop, uv_poll_s * handle, int fd) Line 509	C
 	libwebsockets-test-server-libuv.exe!lws_libuv_accept(lws * wsi, int accept_fd) Line 258	C
 	libwebsockets-test-server-libuv.exe!lws_adopt_socket_vhost(lws_vhost * vh, unsigned int accept_fd) Line 1173	C
 	libwebsockets-test-server-libuv.exe!lws_server_socket_service(lws_context * context, lws * wsi, lws_pollfd * pollfd) Line 1526	C
 	libwebsockets-test-server-libuv.exe!lws_service_fd_tsi(lws_context * context, lws_pollfd * pollfd, int tsi) Line 765	C
 	libwebsockets-test-server-libuv.exe!lws_service_fd(lws_context * context, lws_pollfd * pollfd) Line 1064	C
 	libwebsockets-test-server-libuv.exe!lws_io_cb(uv_poll_s * watcher, int status, int revents) Line 101	C
 	libwebsockets-test-server-libuv.exe!uv__fast_poll_process_poll_req(uv_loop_s * loop, uv_poll_s * handle, uv_req_s * req) Line 214	C
 	libwebsockets-test-server-libuv.exe!uv_process_poll_req(uv_loop_s * loop, uv_poll_s * handle, uv_req_s * req) Line 622	C
 	libwebsockets-test-server-libuv.exe!uv_process_reqs(uv_loop_s * loop) Line 205	C
 	libwebsockets-test-server-libuv.exe!uv_run(uv_loop_s * loop, uv_run_mode mode) Line 391	C
 	libwebsockets-test-server-libuv.exe!lws_libuv_run(const lws_context * context, int tsi) Line 329	C
 	libwebsockets-test-server-libuv.exe!main(int argc, char * * argv) Line 318	C```

**Comment by lucasterra** [comment]
> ... If you pass libuv the wrong type it will blow up.

This might not be specific to libuv, since _test-server-ext-poll_ doesn't work either.

**Comment by lws-team** [comment]
Does the normal test server work okay?

**Comment by lucasterra** [comment]
Yes, normal test server works fine.

**Comment by lws-team** [comment]
I guess this isn't master?  The fix I am remembering is this 9694ebefe8367882c79585e7f1a5656ea2a55ce7 where he uses uv_poll_init_socket()

**Comment by lucasterra** [comment]
It is master :cry: 
Maybe one [uv_poll_init](https://github.com/warmcat/libwebsockets/blob/master/lib/libuv.c#L258) was left behind in lws_libuv_accept

**Comment by lucasterra** [comment]
Yeah. If I change it to ```uv_poll_init_socket``` it won't assert anymore, but the requests will time out:

```
[2016/03/18 21:56:02:0090] DEBUG: fd=260, revents=9
[2016/03/18 21:56:02:0095] DEBUG: accepted new conn  port 20562 on fd=452
[2016/03/18 21:56:02:0095] DEBUG: lws_set_timeout: 0104E5A8: 5 secs
[2016/03/18 21:56:02:0095] DEBUG: lws_libuv_accept: new wsi 0104E5A8
[2016/03/18 21:56:02:0095] DEBUG: insert_wsi_socket_into_fds: 0104E5A8: tsi=0, sock=452, pos-in-fds=1
[2016/03/18 21:56:02:0100] DEBUG: accepted new conn  port 20563 on fd=464
[2016/03/18 21:56:02:0100] DEBUG: lws_set_timeout: 0104F150: 5 secs
[2016/03/18 21:56:02:0100] DEBUG: lws_libuv_accept: new wsi 0104F150
[2016/03/18 21:56:02:0100] DEBUG: insert_wsi_socket_into_fds: 0104F150: tsi=0, sock=464, pos-in-fds=2
[2016/03/18 21:56:02:0105] DEBUG: lws_uv_idle
[2016/03/18 21:56:02:0105] DEBUG: lws_uv_idle: done stop
[2016/03/18 21:56:02:2605] DEBUG: fd=260, revents=9
[2016/03/18 21:56:02:2610] DEBUG: accepted new conn  port 20564 on fd=472
[2016/03/18 21:56:02:2620] DEBUG: lws_set_timeout: 0104F600: 5 secs
[2016/03/18 21:56:02:2625] DEBUG: lws_libuv_accept: new wsi 0104F600
[2016/03/18 21:56:02:2630] DEBUG: insert_wsi_socket_into_fds: 0104F600: tsi=0, sock=472, pos-in-fds=3
[2016/03/18 21:56:02:2630] DEBUG: lws_uv_idle
[2016/03/18 21:56:02:2635] DEBUG: lws_uv_idle: done stop
[2016/03/18 21:56:02:3911] DEBUG: lws_uv_timeout_cb
[2016/03/18 21:56:03:3914] DEBUG: lws_uv_timeout_cb
[2016/03/18 21:56:04:3914] DEBUG: lws_uv_timeout_cb
[2016/03/18 21:56:05:3926] DEBUG: lws_uv_timeout_cb
[2016/03/18 21:56:06:3924] DEBUG: lws_uv_timeout_cb
[2016/03/18 21:56:07:3928] DEBUG: lws_uv_timeout_cb
[2016/03/18 21:56:08:3932] DEBUG: lws_uv_timeout_cb
[2016/03/18 21:56:08:3932] NOTICE: wsi 0104F600: TIMEDOUT WAITING on 3 (did hdr 0, ah 00000000, wl 0, pfd events 0)
[2016/03/18 21:56:08:3949] DEBUG: lws_libuv_io: wsi: 0104F600, flags:0x8000000b
[2016/03/18 21:56:08:3954] DEBUG: lws_libuv_io: wsi: 028E8510, flags:0x5
[2016/03/18 21:56:08:3954] DEBUG: not calling back closed mode=0 state=0
[2016/03/18 21:56:08:3959] DEBUG: lws_close_free_wsi: lws_libuv_closehandle: wsi 0104F600
[2016/03/18 21:56:08:3964] NOTICE: wsi 0104F150: TIMEDOUT WAITING on 3 (did hdr 0, ah 00000000, wl 0, pfd events 0)
[2016/03/18 21:56:08:3964] DEBUG: lws_libuv_io: wsi: 0104F150, flags:0x8000000b
[2016/03/18 21:56:08:3969] DEBUG: lws_libuv_io: wsi: 028E8510, flags:0x5
[2016/03/18 21:56:08:3974] DEBUG: not calling back closed mode=0 state=0
[2016/03/18 21:56:08:3990] DEBUG: lws_close_free_wsi: lws_libuv_closehandle: wsi 0104F150
[2016/03/18 21:56:08:3995] NOTICE: wsi 0104E5A8: TIMEDOUT WAITING on 3 (did hdr 0, ah 00000000, wl 0, pfd events 0)
[2016/03/18 21:56:08:4000] DEBUG: lws_libuv_io: wsi: 0104E5A8, flags:0x8000000b
[2016/03/18 21:56:08:4000] DEBUG: lws_libuv_io: wsi: 028E8510, flags:0x5
[2016/03/18 21:56:08:4005] DEBUG: not calling back closed mode=0 state=0
[2016/03/18 21:56:08:4010] DEBUG: lws_close_free_wsi: lws_libuv_closehandle: wsi 0104E5A8
[2016/03/18 21:56:08:4010] DEBUG: lws_free_wsi: 0104E5A8, remaining wsi 3
[2016/03/18 21:56:08:4015] DEBUG: lws_free_wsi: 0104F150, remaining wsi 2
[2016/03/18 21:56:08:4030] DEBUG: lws_free_wsi: 0104F600, remaining wsi 1
[2016/03/18 21:56:09:3946] DEBUG: lws_uv_timeout_cb
[2016/03/18 21:56:10:3947] DEBUG: lws_uv_timeout_cb
[2016/03/18 21:56:11:3951] DEBUG: lws_uv_timeout_cb
```

**Comment by lws-team** [comment]
Thanks I pushed a patch changing the one in accept as well.

Here's what the log looks like on Linux for wgetting on http

```
...
lwsws[9811]: lws_uv_timeout_cb
lwsws[9811]: lws_uv_timeout_cb
lwsws[9811]: lws_uv_timeout_cb
lwsws[9811]: fd=15, revents=1
lwsws[9811]: accepted new conn  port 42052 on fd=11

lwsws[9811]: Accepted 0x1386a50 to tsi 0
lwsws[9811]: lws_adopt_socket_vhost: new wsi 0x1386a50, sockfd 11
lwsws[9811]: lws_set_timeout: 0x1386a50: 20 secs
lwsws[9811]: lws_libuv_accept: new wsi 0x1386a50
lwsws[9811]: insert_wsi_socket_into_fds: 0x1386a50: tsi=0, sock=11, pos-in-fds=3
lwsws[9811]: lws_libuv_io: wsi: 0x1386a50, flags:0x5
lwsws[9811]: lws_uv_idle
lwsws[9811]: lws_uv_idle: done stop
lwsws[9811]: fd=11, revents=1
lwsws[9811]: lws_header_table_attach: wsi 0x1386a50: ah (nil) (tsi 0)
lwsws[9811]: lws_libuv_io: wsi: 0x1386a50, flags:0x5
lwsws[9811]: lws_header_table_attach: wsi 0x1386a50: ah 0x1342d30: count 1 (on exit)
lwsws[9811]: lws_server_socket_service: 0x1386a50: rxpos:0 rxlen:0
lwsws[9811]: lws_server_socket_service: wsi 0x1386a50, ah->rxlen = 146
lwsws[9811]: lws_read: incoming len 146  state 0
lwsws[9811]: issuing 146 bytes to parser
lwsws[9811]: WSI_TOKEN_NAME_PART 'G' (mode=0)
lwsws[9811]: WSI_TOKEN_NAME_PART 'E' (mode=0)
lwsws[9811]: WSI_TOKEN_NAME_PART 'T' (mode=0)
lwsws[9811]: WSI_TOKEN_NAME_PART ' ' (mode=0)
lwsws[9811]: known hdr 0
...
```

I added a blank line where it deviates (this is with -d65535)

**Comment by lws-team** [comment]
Hm it looks like no ah gets attached.... the POLLIN processing is defeated until an ah is attached, the connections should sit there until an ah is available from the pool.  Is there anything like this in the log before?

```
 lws_header_table_attach: wsi 0x1386a50: ah 0x1342d30: count 1 (on exit)
```

**Comment by lucasterra** [comment]
The only ah that shows up on my logs is:

[2016/03/18 22:34:45:2832] NOTICE: wsi 0249F058: TIMEDOUT WAITING on 3 (did hdr 0, ah 00000000, wl 0, pfd events 0)

**Comment by lucasterra** [comment]
I added a "print callback reason" to protocol_http_cb, and got this. Not sure if it helps:

```
[2016/03/18 22:46:26:5516] DEBUG: fd=252, revents=9
[2016/03/18 22:46:26:5521] DEBUG: accepted new conn  port 28637 on fd=296
[2016/03/18 22:46:26:5526] NOTICE: callback_http reason: 'LWS_CALLBACK_FILTER_NETWORK_CONNECTION'
[2016/03/18 22:46:26:5526] INFO: Accepted 024A0CF0 to tsi 0
[2016/03/18 22:46:26:5531] NOTICE: callback_http reason: 'LWS_CALLBACK_WSI_CREATE'
[2016/03/18 22:46:26:5531] INFO: lws_adopt_socket_vhost: new wsi 024A0CF0, sockfd 296
[2016/03/18 22:46:26:5531] DEBUG: lws_set_timeout: 024A0CF0: 5 secs
[2016/03/18 22:46:26:5531] NOTICE: callback_http reason: 'LWS_CALLBACK_SERVER_NEW_CLIENT_INSTANTIATED'
[2016/03/18 22:46:26:5531] DEBUG: lws_libuv_accept: new wsi 024A0CF0
[2016/03/18 22:46:26:5545] DEBUG: insert_wsi_socket_into_fds: 024A0CF0: tsi=0, sock=296, pos-in-fds=1
[2016/03/18 22:46:26:5545] NOTICE: callback_http reason: 'LWS_CALLBACK_LOCK_POLL'
[2016/03/18 22:46:26:5569] NOTICE: callback_http reason: 'LWS_CALLBACK_ADD_POLL_FD'
[2016/03/18 22:46:26:5574] NOTICE: callback_http reason: 'LWS_CALLBACK_UNLOCK_POLL'
[2016/03/18 22:46:26:5574] DEBUG: lws_uv_idle
[2016/03/18 22:46:26:5579] DEBUG: lws_uv_idle: done stop
[2016/03/18 22:46:27:0122] DEBUG: lws_uv_timeout_cb
[2016/03/18 22:46:28:0123] DEBUG: lws_uv_timeout_cb
[2016/03/18 22:46:29:0127] DEBUG: lws_uv_timeout_cb
[2016/03/18 22:46:30:0129] DEBUG: lws_uv_timeout_cb
[2016/03/18 22:46:31:0129] DEBUG: lws_uv_timeout_cb
[2016/03/18 22:46:32:0130] DEBUG: lws_uv_timeout_cb
[2016/03/18 22:46:32:0130] NOTICE: wsi 024A0CF0: TIMEDOUT WAITING on 3 (did hdr 0, ah 00000000, wl 0, pfd events 0)
[2016/03/18 22:46:32:0135] INFO: lws_close_free_wsi: real just_kill_connection: 024A0CF0 (sockfd 296)
[2016/03/18 22:46:32:0135] NOTICE: callback_http reason: 'LWS_CALLBACK_LOCK_POLL'
[2016/03/18 22:46:32:0135] DEBUG: lws_libuv_io: wsi: 024A0CF0, flags:0x8000000b
[2016/03/18 22:46:32:0148] INFO: remove_wsi_socket_from_fds: wsi=024A0CF0, sock=296, fds pos=1, end guy pos=2, endfd=464
[2016/03/18 22:46:32:0153] NOTICE: callback_http reason: 'LWS_CALLBACK_DEL_POLL_FD'
[2016/03/18 22:46:32:0153] NOTICE: callback_http reason: 'LWS_CALLBACK_CHANGE_MODE_POLL_FD'
[2016/03/18 22:46:32:0158] DEBUG: lws_libuv_io: wsi: 02729518, flags:0x5
[2016/03/18 22:46:32:0163] NOTICE: callback_http reason: 'LWS_CALLBACK_UNLOCK_POLL'
[2016/03/18 22:46:32:0163] DEBUG: not calling back closed mode=0 state=0
[2016/03/18 22:46:32:0193] DEBUG: lws_close_free_wsi: lws_libuv_closehandle: wsi 024A0CF0
[2016/03/18 22:46:32:0198] NOTICE: callback_http reason: 'LWS_CALLBACK_WSI_DESTROY'
[2016/03/18 22:46:32:0203] INFO: lws_header_table_detach: wsi 024A0CF0: ah 00000000 (tsi=0, count = 0)
[2016/03/18 22:46:32:0208] DEBUG: lws_free_wsi: 024A0CF0, remaining wsi 1
```

**Comment by lws-team** [comment]
Does this do anything?

```
diff --git a/lib/server.c b/lib/server.c
index 41e6339..f1c29e5 100644
--- a/lib/server.c
+++ b/lib/server.c
@@ -1200,6 +1200,9 @@ lws_adopt_socket_vhost(struct lws_vhost *vh, lws_sockfd_type accept_fd)
                }
        }
 
+       if (!lws_header_table_attach(new_wsi, 0))
+               lwsl_debug("Attached ah immediately\n");
+
        return new_wsi;
 
 fail:
@@ -1279,7 +1282,7 @@ lws_adopt_socket_readbuf(struct lws_context *context, lws_sockfd_type accept_fd,
         * readbuf data to wsi or ah yet, and we will do it next if we get
         * the ah.
         */
-       if (!lws_header_table_attach(wsi, 0)) {
+       if (wsi->u.hdr.ah || !lws_header_table_attach(wsi, 0)) {
                ah = wsi->u.hdr.ah;
                memcpy(ah->rx, readbuf, len);
                ah->rxpos = 0;
```

**Comment by lucasterra** [comment]
Yeap! That seems to fix the http protocol, but WebSocket still doesn't work:

Chrome Console:
```WebSocket connection to 'ws://localhost:7681/xxx' failed: Error during WebSocket handshake: net::ERR_CONNECTION_RESET```

**Comment by lucasterra** [comment]
Thanks for the help so far, I gotta head to bed. I'll be back tomorrow in the morning.

**Comment by lws-team** [comment]
Well... it's progress... the patch doesn't seem to break Linux so I'll apply it.

Have a nice evening in the meanwhile.

**Comment by lws-team** [comment]
... and tomorrow please send the full (-d65535) logs around the connection that drops from lws side.

**Comment by lucasterra** [comment]
Update: sometimes it works, sometimes it doesn't.
1. It loads http://localhost:7681 and connects to the websocket (but the latency is really high, maybe because of all the logging -d65535)
2. Restart the server
3. Try opening reloading http://localhost:7681 and it won't connect anymore.
4. If you don't restart the server, and keep trying to connect from time to time it will connect eventually. It's really random.

Here is a log when it fails:
```
[2016/03/19 12:48:49:6298] DEBUG: fd=264, revents=9
[2016/03/19 12:48:49:6303] DEBUG: accepted new conn  port 35980 on fd=480
[2016/03/19 12:48:49:6308] NOTICE: callback_http reason: 'LWS_CALLBACK_FILTER_NETWORK_CONNECTION'
[2016/03/19 12:48:49:6313] INFO: Accepted 00DA0CA8 to tsi 0
[2016/03/19 12:48:49:6313] NOTICE: callback_http reason: 'LWS_CALLBACK_WSI_CREATE'
[2016/03/19 12:48:49:6318] INFO: lws_adopt_socket_vhost: new wsi 00DA0CA8, sockfd 480
[2016/03/19 12:48:49:6323] DEBUG: lws_set_timeout: 00DA0CA8: 5 secs
[2016/03/19 12:48:49:6328] NOTICE: callback_http reason: 'LWS_CALLBACK_SERVER_NEW_CLIENT_INSTANTIATED'
[2016/03/19 12:48:49:6333] DEBUG: lws_libuv_accept: new wsi 00DA0CA8
[2016/03/19 12:48:49:6333] DEBUG: insert_wsi_socket_into_fds: 00DA0CA8: tsi=0, sock=480, pos-in-fds=2
[2016/03/19 12:48:49:6338] NOTICE: callback_http reason: 'LWS_CALLBACK_LOCK_POLL'
[2016/03/19 12:48:49:6357] NOTICE: callback_http reason: 'LWS_CALLBACK_ADD_POLL_FD'
[2016/03/19 12:48:49:6362] NOTICE: callback_http reason: 'LWS_CALLBACK_UNLOCK_POLL'
[2016/03/19 12:48:49:6362] INFO: lws_header_table_attach: wsi 00DA0CA8: ah 00000000 (tsi 0)
[2016/03/19 12:48:49:6367] INFO: lws_header_table_attach: adding 00DA0CA8 to ah waiting list
[2016/03/19 12:48:49:6372] NOTICE: callback_http reason: 'LWS_CALLBACK_CHANGE_MODE_POLL_FD'
[2016/03/19 12:48:49:6372] DEBUG: lws_libuv_io: wsi: 00DA0CA8, flags:0x9
[2016/03/19 12:48:49:6377] DEBUG: lws_uv_idle
[2016/03/19 12:48:49:6382] DEBUG: lws_uv_idle: done stop
[2016/03/19 12:48:50:5930] DEBUG: lws_uv_timeout_cb
[2016/03/19 12:48:51:5935] DEBUG: lws_uv_timeout_cb
[2016/03/19 12:48:52:5936] DEBUG: lws_uv_timeout_cb
[2016/03/19 12:48:53:5945] DEBUG: lws_uv_timeout_cb
[2016/03/19 12:48:54:5948] DEBUG: lws_uv_timeout_cb
[2016/03/19 12:48:55:5954] DEBUG: lws_uv_timeout_cb
[2016/03/19 12:48:55:5954] NOTICE: wsi 00DA0CA8: TIMEDOUT WAITING on 3 (did hdr 0, ah 00000000, wl 1, pfd events 0)
[2016/03/19 12:48:55:5958] INFO: lws_close_free_wsi: real just_kill_connection: 00DA0CA8 (sockfd 480)
[2016/03/19 12:48:55:5963] NOTICE: callback_http reason: 'LWS_CALLBACK_LOCK_POLL'
[2016/03/19 12:48:55:5970] DEBUG: lws_libuv_io: wsi: 00DA0CA8, flags:0x8000000b
[2016/03/19 12:48:55:5970] INFO: remove_wsi_socket_from_fds: wsi=00DA0CA8, sock=480, fds pos=2, end guy pos=3, endfd=484
[2016/03/19 12:48:55:5975] NOTICE: callback_http reason: 'LWS_CALLBACK_DEL_POLL_FD'
[2016/03/19 12:48:55:5975] NOTICE: callback_http reason: 'LWS_CALLBACK_CHANGE_MODE_POLL_FD'
[2016/03/19 12:48:55:5980] DEBUG: lws_libuv_io: wsi: 02678510, flags:0x5
[2016/03/19 12:48:55:5985] NOTICE: callback_http reason: 'LWS_CALLBACK_UNLOCK_POLL'
[2016/03/19 12:48:55:5985] DEBUG: not calling back closed mode=0 state=0
[2016/03/19 12:48:55:5990] DEBUG: lws_close_free_wsi: lws_libuv_closehandle: wsi 00DA0CA8
[2016/03/19 12:48:55:6009] NOTICE: callback_http reason: 'LWS_CALLBACK_WSI_DESTROY'
[2016/03/19 12:48:55:6014] INFO: lws_header_table_detach: wsi 00DA0CA8: ah 00000000 (tsi=0, count = 1)
[2016/03/19 12:48:55:6019] INFO: lws_header_table_detach: wsi 00DA0CA8, remv wait
[2016/03/19 12:48:55:6019] DEBUG: lws_free_wsi: 00DA0CA8, remaining wsi 2
[2016/03/19 12:48:56:5957] DEBUG: lws_uv_timeout_cb
```

**Comment by lws-team** [comment]
It sounds like the only problem is the latency, if it takes you over the 5s timeout the connection is dropped just from that.  If you increase the timeout (you can do it in info struct at context creation time) that should become clear.

Also in the info struct, if the ah pool is set to 1, try increasing it to 8.

**Comment by lucasterra** [comment]
Yeah. Changing ```max_http_header_pool``` (that's what you meant by ah pool, right?) to 8 fixed it.
Is there any side effects this could cause?

**Comment by lucasterra** [comment]
Anyways, there might still be a bug somewhere, because it's taking too long to release the buffer back to the pool (when max_http_header_pool is set to 1).

**Comment by lws-team** [comment]
Nice.

There's no side effects from that, but it should work properly with just one in the pool, it's left like that in the test server as a test of the ah lifecycle and waiting list stuff.  It means there's some other problem with ah getting left attached... since it works OK with normal poll loop that's kind of suspicious.

Basically the connections must acquire an ah from the pool to process http headers, if all the ah are busy, the connection joins a waiting list and waits until one becomes available.  When it does, it's attached and the headers are read and processed.

In this case your log showed that when the connection timed out he was on the ah waiting list, ie he spent his whole life waiting for an ah.

The ah structs are a few KB so allocating 8 or 16 or whatever at context time is not so burdensome... the pool was created because before they were malloc'd on demand and there was no real management of the memory involved in the case of eg, a connection storm.

**Comment by lucasterra** [comment]
Just to provide more info...

If I connect to the websocket from an already loaded page it works fine. i.e.:
```javascript
var ws = new WebSocket('ws://localhost:7681/xxx', "dumb-increment-protocol");
ws.onopen = function(e) {
  console.log('connected')
};
```

If I refresh the page (meaning it has to use the http pool to serve the static html content), then the websockets won't connect.

**Comment by lws-team** [comment]
Both cases need to process http headers and need an ah, the ws connection also starts out as normal http and does an upgrade in his http headers.

I guess the difference is the ws connection can reuse a keep-alive http connection that loaded the page, whereas F5 makes a new connection... if the subsequent keep-alive action comes before the completion of the first transaction (ie, is pipelined) lws will keep the ah from the first one because it sees it has another transaction buffered to do right away.  Otherwise at the end of each keepalive transaction it gives up the ah until something new comes to deal with on the keepalive connection.

The interesting thing would be understanding who got the (only) ah and what happened on that connection that he ended up keeping it.  I'm not sure why libuv would affect this stuff since it should be unrelated to the event loop.

**Comment by lucasterra** [comment]
But refreshing makes it open like 5 connections to the server:
![connections](https://cloud.githubusercontent.com/assets/441058/14658359/87954140-0669-11e6-8eb7-5b357a730a1e.png)

Where direct connecting opens only 1. Now... who has the ah?

**Comment by lws-team** [comment]
Each will have had the ah in turn, while it was in the headers phase, otherwise it can't get past that.  After the connections are set up and any remaining http/1.1 connections are idle in keepalive, nobody should have the ah any more, it should be back in the pool.

Only in the case we just did an http/1.1 transaction and the next is already pipelined and ready to process, should we keep the ah.

**Comment by lws-team** [comment]
Thanks for the debugging effort... I'm going to close this since the original problem is solved.  Default ah pool is actually 16 (if you leave it at 0 in the info) so that seems reasonable to work around if otherwise OK.

It's a bit suspicious there's a problem with sticky ah as to where it's coming from, I can't see how either uv can create that situation if otherwise working smoothly or windows plat stuff could do it.  So it might reflect some general corner case.

If there's new info please reopen or other problems please make a new issue.  Thanks again for the debugging.
